File: cpp-icestorm/subscriber.cpp
#include <Ice/Application.h>
#include <IceStorm/IceStorm.h>
#include <IceUtil/UUID.h>
#include "Printer.h"

using namespace std;
using namespace Ice;
using namespace IceStorm;
using namespace Example;


class PrinterI : public Printer {
public:
  void write(const string& message, const Current& current) {
    cout << "Event received: " << message << endl;
  }
};


class Subscriber : public Application {
  TopicManagerPrx get_topic_manager() {
    string key = "IceStorm.TopicManager.Proxy";
    ObjectPrx proxy = communicator()->propertyToProxy(key);
    if (!proxy) {
      cerr << "property " << key << " not set." << endl;
      return 0;
    }

    cout << "Using IceStorm in: '" << key << "' " << endl;
    return TopicManagerPrx::checkedCast(proxy);
  }

public:
  virtual int run(int argc, char* argv[]) {
    TopicManagerPrx topic_mgr = get_topic_manager();
    if (!topic_mgr) {
      cerr << appName() << ": invalid proxy" << endl;
      return EXIT_FAILURE;
    }

    ObjectPtr servant = new PrinterI;
    ObjectAdapterPtr adapter = \
      communicator()->createObjectAdapter("PrinterAdapter");
    ObjectPrx subscriber = adapter->addWithUUID(servant);

    string topic_name = "PrinterTopic";
    TopicPrx topic;
    try {
      topic = topic_mgr->retrieve(topic_name);
    } catch(const NoSuchTopic& e) {
      topic = topic_mgr->create(topic_name);
    }

    topic->subscribeAndGetPublisher(QoS(), subscriber);
    cout << "Waiting events... " << subscriber << endl;

    adapter->activate();
    shutdownOnInterrupt();
    communicator()->waitForShutdown();

    topic->unsubscribe(subscriber);

    return EXIT_SUCCESS;
  }
};

int main(int argc, char* argv[]) {
  Subscriber app;
  return app.main(argc, argv);
}

File: java-icestorm/Subscriber.java
import java.util.HashMap;
import Ice.*;
import IceStorm.*;
import Example.*;

public class Subscriber extends Application {
  public class PrinterI extends _PrinterDisp {
    public void write(String message, Ice.Current current) {
      System.out.println(String.format("Event received: %s", message));
    }
  }

  public int run(String[] args) {
    String key = "IceStorm.TopicManager.Proxy";
    ObjectPrx prx = communicator().propertyToProxy(key);
    TopicManagerPrx manager = TopicManagerPrxHelper.checkedCast(prx);

    if (manager == null) {
      System.err.println("invalid proxy");
      return 1;
    }

    PrinterI servant = new PrinterI();
    ObjectAdapter adapter =
	   communicator().createObjectAdapter("PrinterAdapter");
    ObjectPrx subscriber = adapter.addWithUUID(servant);

    String topic_name = "PrinterTopic";
    TopicPrx topic;
    try {
      topic = manager.retrieve(topic_name);
    } catch(NoSuchTopic e) {
      try {
        topic = manager.create(topic_name);
      } catch(TopicExists ex) {
        System.err.println("Temporary failure, try again.");
        return 1;
      }
    }

    try {
      topic.subscribeAndGetPublisher(null, subscriber);
    } catch(InvalidSubscriber e) {
      e.printStackTrace();
      return 1;
    } catch(AlreadySubscribed e) {
      e.printStackTrace();
      return 1;
    } catch(BadQoS e) {
      e.printStackTrace();
      return 1;
    }

    System.out.println("Waiting events... ");

    adapter.activate();
    shutdownOnInterrupt();
    communicator().waitForShutdown();

    topic.unsubscribe(subscriber);

    return 0;
  }

  public static void main(String[] args) {
    Subscriber app = new Subscriber();
    int status = app.main("Subscriber", args);
    System.exit(status);
  }
}
